NaN can be used as a key in Map

Akirathan opened the pull request:
Fixes #5833 #6065 

### Pull Request Description

`Number.nan` can be used as a key in `Map`. This PR basically implements the support for [JavaScript's Same Value Zero Equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality) so that `Number.nan` can be used as a key in `Map`.

### Important Notes
- For NaN, it holds that `Meta.is_same_object Number.nan Number.nan`, and `Number.nan != Number.nan` - inspired by JS spec.
- `Meta.is_same_object x y` implies `Any.== x y`, except for `Number.nan`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

Commit:
Check for Meta.is_same_object in Any.==

Commit:
Add some tests for Vector.distinct

mwu-tow commented:
@Akirathan 
> `Number.nan` can be used as a key in `Map`.

Why do we want this? For me `NaN` is the value that is not equal to itself. That by definition kind of prevents it from being useful as a key in the map.
Even if we wanted this, I'd rather expect some kind of byte-representation-comparison rather than just `NaN==NaN` assumption.  There are different values (byte-wise) that are considered to be a 'NaN'. Do we want to disregard the differences?

Commit:
`Meta.is_same_object NaN NaN` returns True

Commit:
`Meta.is_same_object NaN NaN` returns True

Akirathan commented:
> @Akirathan
> 
> > `Number.nan` can be used as a key in `Map`.
> 
> Why do we want this? For me `NaN` is the value that is not equal to itself. That by definition kind of prevents it from being useful as a key in the map. Even if we wanted this, I'd rather expect some kind of byte-representation-comparison rather than just `NaN==NaN` assumption. There are different values (byte-wise) that are considered to be a 'NaN'. Do we want to disregard the differences?

As noted in the issue description #5833, we want to comply with [JavaScript's same-value-zero equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality). The reason for that is because we want *every* value in Enso, including polyglot values, to have some hash code, and thus be used as a key in the map.

Commit:
Add Map tests for various incomparable keys

Commit:
Test that `Meta.is_same_object` complies with "Same Value Zero" equality.

Spec at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality

Commit:
Revert "`Meta.is_same_object NaN NaN` returns True"

This reverts commit 32d81f46c0d54d533e3404c98dcffaa1dbb00230.

Commit:
Add docs to Map

Commit:
Merge branch 'develop' into wip/akirathan/nan-in-map-5833

# Conflicts:
#	distribution/lib/Standard/Base/0.0.0-dev/src/Any.enso

Commit:
Any.== checks for Meta.is_same_object (#6065)

Commit:
Remove "should support vector with NaNs as keys".

This is not according to the JS spec.

Commit:
Fix typos in tests

Commit:
bench_download tool: fail if there are no job reports

Akirathan commented on test/Tests/src/Data/Map_Spec.enso:
```diff
@@ -143,6 +168,33 @@ spec =
+            Map.singleton (Map.singleton Number.nan 1) 42 . keys . at 0 . get Number.nan . should_equal 1
+            Map.singleton (Map.singleton Number.nan 1) 42 . at (Map.singleton Number.nan 1) . should_equal 42
+
+        Test.specify "should support atoms with custom comparators that violate reflexivity as keys" <|
```
This test was copied from the issue description

Akirathan commented on test/Tests/src/Semantic/Meta_Spec.enso:
```diff
@@ -281,13 +281,18 @@ spec =
         Test.specify "type_of Nothing is Nothing" <|
             Meta.type_of Nothing . should_equal Nothing
 
+        Test.specify "NaN and NaN should be the same object" <|
```
This is exactly how JS works, according to the spec in [Same value zero equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality)

Commit:
Use IsSameObject instead of reference comparison

Commit:
Merge branch 'develop' into wip/akirathan/nan-in-map-5833

JaroslavTulach commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Map.enso:
```diff
@@ -15,6 +15,12 @@ from project import Error, Nothing, Any, Panic
    different languages, etc.
 
+   For keys that are not reflexive, like `Number.nan`, "Same Value Equality", from JavaScript
+   specification in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality,
```
I think we can use Markdown format in the comments. Then I recommend to use
```markdown
[Same Value Equality](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#same-value-zero_equality)
```
e.g. hyperlink text without the URL being visible.

Akirathan replied:
Done in fc5aace5a9cd0cc3a3a00ba30dfd2fcb1199b96e

Commit:
Small fixes after merge

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:
```diff
@@ -214,43 +209,43 @@ boolean equalsStrings(Object selfString, Object otherString,
+      Atom other,
+      @Cached EqualsAtomNode equalsNode,
+      @Cached IsSameObjectNode isSameObjectNode) {
+    return isSameObjectNode.execute(self, other) || equalsNode.execute(self, other);
```
This is the fix for #6065, right?

Akirathan replied:
Yes, #6065 is addressed in these commits:
- 35074849afa320cac15d866b9c7f3921ef3f2bc4
- 55ceaba9b1ff2a0cf63fec5b11aabbad834c0e06

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsSameObjectNode.java:
```diff
@@ -23,6 +25,15 @@ public static IsSameObjectNode build() {
 
+  @Specialization
+  boolean isSameDouble(double left, double right) {
+    if (Double.isNaN(left) && Double.isNaN(right)) {
```
This is the handling of `NaN` as being _"same"_.

JaroslavTulach commented on test/Tests/src/Semantic/Meta_Spec.enso:
```diff
@@ -281,13 +281,18 @@ spec =
 
+        Test.specify "NaN and NaN should be the same object" <|
+            Meta.is_same_object Number.nan Number.nan . should_be_true
+            (Number.nan == Number.nan) . should_be_false
```
`nan != nan`, but `Meta.is_same_object` gives `true`. OK.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsSameObjectNode.java:
```diff
@@ -23,6 +25,15 @@ public static IsSameObjectNode build() {
 
+  @Specialization
+  boolean isSameDouble(double left, double right) {
+    if (Double.isNaN(left) && Double.isNaN(right)) {
+      return true;
+    } else {
+      return left == right;
+    }
```
```suggestion
    return (Double.isNaN(left) && Double.isNaN(right)) || (left == right);
```
personal preference feel free to ignore :p

hubertp commented on tools/performance/engine-benchmarks/bench_download.py:
```diff
@@ -706,14 +706,20 @@ def commit_to_str(commit: Commit) -> str:
             if job_report:
                 job_reports.append(job_report)
         logging.debug(f"Got {len(job_reports)} job reports")
+        if len(job_reports) == 0:
```
That seems unrelated?

Akirathan replied:
Yes. I added it because just yesterday I uploaded new benchmark results, and stepped into this issue, so I committed this tiny change along in this PR.

hubertp reviewed (approved):
Do we want/need any tests for polyglot handling of NaNs?

Commit:
Use markdown in docs

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:
```diff
@@ -32,9 +33,9 @@
-      references point to the same object on the heap.
-      """
-)
+      references point to the same object on the heap. However `Any.==` implies
+      `Meta.is_same_object` for all object with the exception of `Number.nan`.
+      """)
```
This does not seem true, or am I misunderstanding?

If I have two separate instances of an object, they may be `==` but will not be `is_same_object`, right?

Shouldn't this condition be contraposed?

If I have `is_same_object x y` then I'm guranteed that `x == y`, as long as they are not `NaN`.

Akirathan replied:
You are right. I wrote the implication in the wrong direction. Fixed in d65e5e9f38a7b8b77e9673447af301744ff2d0a7.

Commit:
Fix implication in docs

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java:
```diff
@@ -44,11 +45,7 @@ public static EqualsNode build() {
-   * Primitive values
-   **/
-
-
+  /** Primitive values */
```
This used to be a 'section' comment, but now reformatting seems to have attached it to the `equalsBoolBool` method.

That doesn't seem right.

I think we should either keep these as 'sections' or remove them. Being attached to the first method is just awkward.

Akirathan replied:
You are right. I will remove it. Recently, I enabled a plugin in my IDE that does automatic reformatting on commits and this is the result.

radeusgd commented on test/Tests/src/Data/Vector_Spec.enso:
```diff
@@ -654,8 +654,23 @@ type_spec name alter = Test.group name <|
         alter [1, 1.0, 2, 2.0] . distinct . should_equal [1, 2]
         alter [] . distinct . should_equal []
 
+    Test.specify "should be able to handle distinct on different primitive values" <|
+        alter [1, "a"] . distinct . should_equal [1, "a"]
+        alter ["a", 1] . distinct . should_equal ["a", 1]
+        alter [Nothing, Nothing] . distinct . should_equal [Nothing]
+        alter [Number.nan, Number.nan] . distinct . to_text . should_equal "[NaN]"
+        alter [Nothing, Number.nan, Nothing, Number.nan] . distinct . to_text . should_equal "[Nothing, NaN]"
+        alter [1, Nothing] . distinct . should_equal [1, Nothing]
+        alter [Nothing, 1, Nothing] . distinct . should_equal [Nothing, 1]
+        alter [1, Number.nan] . distinct . to_text . should_equal "[1, NaN]"
+        alter [Number.nan, 1, Number.nan] . distinct . to_text . should_equal "[NaN, 1]"
+        alter [1, Number.nan, 1] . distinct . to_text . should_equal "[1, NaN]"
```
How does `distinct` work with other non-reflexive values, like `My_Nan`? Can we get a test for it too?

Akirathan replied:
Added in 6bae7283fe1ce8276de282dcee37ef45abaa7d89

radeusgd reviewed (approved):
Looks good, I'd just appreciate a test for `My_Nan` in `distinct`.

This also should at some point work consistently in `Table.aggregate` - `Aggregate_Column.Group_By`. I guess no need to add these tests now, but maybe we should add a note to #6292 to not forget about it? It is quite a brittle feature and easy to get wrong so it will be good to have tests for this case once available.

Commit:
Add more incomparable tests to Vector_Spec
